File: Data Structures & Algorithms/search-2d-matrix/submission-1.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m = len(matrix)
        n = len(matrix[0])

        l, r = 0, m
        while l+1 < r:
            mid = (l+r) // 2
            if matrix[mid][0] == target:
                return True
            elif matrix[mid][0] < target:
                l = mid
            else:
                r = mid
        
        l1, r1 = 0, n
        while l1 + 1 < r1:
            mid = (l1+r1) // 2
            if matrix[l][mid] == target:
                return True
            elif matrix[l][mid] < target:
                l1 = mid
            else:
                r1 = mid
        return matrix[l][l1] == target
        

File: Data Structures & Algorithms/minimum-size-subarray-sum/submission-1.cpp
class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums) {
        vector<int> pSum(nums.size()+1);

        for(int i = 1; i <= nums.size(); i++){
            pSum[i] = pSum[i-1] + nums[i-1];
        }

        if(pSum[nums.size()] < target) return 0;

        int left = 0, right = nums.size();
        while(left+1 < right){
            int mid = (left+right)/2;

            bool canDo = false;
            for(int i = 0; i < pSum.size()-mid; i++){
                if(pSum[i+mid]-pSum[i] >= target){
                    canDo = true;
                    break;
                }
            }

            if(canDo) right = mid;
            else left = mid;
        }

        return right;
    }
};